### PR TITLE
Rewrite WebUtility.HtmlEntities to minimize array/substring allocations

### PR DESCRIPTION
**Pull request by jamesqo**
I was browsing the `WebUtility` source code recently and noticed there a huge static array allocation that was never actually used. In addition to that, all of the strings in the array were being `Substring`'d in a loop to initialize a dictionary, which added even more unnecessary allocations.

I [rewrote](https://gist.github.com/jamesqo/dda32f6e7a4154053862) the code so it straight-up initializes the dictionary as a literal, using C# 6's new dictionary initializers feature. The new code also passes in a predefined capacity to the dictionary, so there's no resizing taking place when elements are added.

### CODE REVIEW

**Comment by davidsh** [comment]
@stephentoub PTAL

File: src/System.Runtime.Extensions/src/System/Net/WebUtility.cs [hunk 1]
```diff
@@ -678,276 +678,266 @@ private static class HtmlEntities
...
-
-            private static LowLevelDictionary<string, char> GenerateLookupTable()
-            {
-                // e[0] is unicode char, e[1] is '-', e[2+] is entity string
+            private const int Count = 253;
 
-                LowLevelDictionary<string, char> lookupTable = new LowLevelDictionary<string, char>(StringComparer.Ordinal);
-                foreach (string e in s_entitiesList)
+            // maps entity strings => unicode chars
+            private static LowLevelDictionary<string, char> s_lookupTable =
```
Comment (justinvp) [inline src/System.Runtime.Extensions/src/System/Net/WebUtility.cs]: I realize it wasn't there previously, but might as well add `readonly` while making changes here.

Comment (jamesqo) [inline src/System.Runtime.Extensions/src/System/Net/WebUtility.cs]: @justinvp Good point; I've fixed that.

File: src/System.Runtime.Extensions/src/System/Net/WebUtility.cs [hunk 2]
```diff
@@ -678,276 +678,266 @@ private static class HtmlEntities
...
-                "\x2665-hearts",
-                "\x2666-diams",
-            };
-
-            private static LowLevelDictionary<string, char> s_lookupTable = GenerateLookupTable();
-
-            private static LowLevelDictionary<string, char> GenerateLookupTable()
-            {
-                // e[0] is unicode char, e[1] is '-', e[2+] is entity string
+            private const int Count = 253;
```
Comment (justinvp) [inline src/System.Runtime.Extensions/src/System/Net/WebUtility.cs]: Consider adding a `DEBUG`-only assert to make sure the constant matches the actual count added, something like:

```csharp
#if DEBUG
        static HtmlEntities()
        {
            // Make sure the initial capacity for s_lookupTable is correct
            Debug.Assert(s_lookupTable.Count == Count, "s_lookupTable's initial capacity is incorrect.");
        }
#endif
```

Example:
https://github.com/dotnet/corefx/blob/ac67ffac987d0c27236c4a6cf1255c2bcbc7fe7d/src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCharClass.cs#L418-L423

Comment (jamesqo) [inline src/System.Runtime.Extensions/src/System/Net/WebUtility.cs]: @justinvp Done.

**Comment by stephentoub** [comment]
LGTM.  And just to double-check I verified via code that the new approach yields a dictionary with exactly the same contents as the old.

**Comment by jamesqo** [comment]
@stephentoub Awesome, thanks!

**Pull request merged**